File: C/labs/lab10/LL.c
// syntax from C
// functions, conditionals, loops
// structs, pointers
// #define, #if/#endif, #include


// Linked lists
// A way to store lists without pre-allocating blocks of memory

// // [2, , ,1, , , ]
//            X
//           ^\
//          /  |
// A->B->C  D->E->F

#include <stdio.h>
#include <stdlib.h>

typedef struct node{
    char *data;
    struct node *next;
} node;
//      10
//     n10
//    /   \
//   1    5
// head   n2

// head->A->B
void create_LL(node **p_head, char *data, int size)
{
    if(size == 0){
        *p_head = NULL;
        return;
    }
    node *cur = (node *)malloc(sizeof(node));
    cur->data = data[0];
    *p_head = cur;
    for(int i = 1; i < size; i++){
        cur->next = (node *)malloc(sizeof(node));
        cur->next->data = data[i];
        cur = cur->next;
    }
    cur->next = NULL;
}
 


// typedef struct node{
//     int data;
//     struct node *next;
// } node; 
void print_LL(node *head)
{
    // head is always the (address of) node we're currently printing
    while(head != NULL){
        printf("%d -> ", head->data);
        head = head->next;
    }
}

void insert_LL(node **p_head, int ind, int num)
{
    node *new = (node *)malloc(sizeof(node));
    new->data = num;
    // insert num at index ind
    if(ind == 0){
        new->next = *p_head;
        *p_head = new;
    }
    else{
        // go through the linked list until node i-1
        // insert the new node after i-1
        node *cur = *p_head;
        for(int i = 0; i < ind - 1; i++){
            cur = cur->next;
        }
        new->next = cur->next;
        cur->next = new;
    }

}

File: C/labs/lab10/LL.py
# Implement a linked list class in Python
# Use it to implement a stack with O(1) push and pop
           
# head->n1-> n2->n3....nk


# head is node*
# head->next is node*
# node *cur
# cur = cur->next

class LL:
    def __init__(self):
        self.head = None

    def insert(self, loc, element):
        new_node = Node(element)  # data: element, next: None
        if loc == 0:
            new_node.next = self.head
            self.head = new_node
        else:
            cur = self.head
            for i in range(loc-1):
                cur = cur.next
            new_node.next = cur.next
            cur.next = new_node
        
    def delete(self, loc):
        if loc == 0:
            self.head = self.head.next 
        else:
            cur = self.head
            for i in range(loc-1):
                cur = cur.next
            # cur is the node before the node that we want to delete
            cur.next = cur.next.next


class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class Stack:
    def __init__(self):
        self.stack = LL() # self.stack is not a linked list with head==None

    # push: insert at 0
    # pop: get 0-th element, and then delete 0th element
    def push(self, elem):
        self.stack.insert(0, elem)

    def pop(self, elem):
        cur_node = self.stack.head
        self.stack.delete(0)
        return cur_node.data

n = Node(52)
# n.data is 52, n.next is None
#n.a9834759432875 = 12